### ISSUE TITLE
Calling .Result on a durable task can cause the orchestrator function to hang

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#201]
Here is an example where someone can run into this unintentionally, like i did:

```csharp
[FunctionName(nameof(MapReduceOrchestration))]
public static async Task<object> MapReduceOrchestration(
    [OrchestrationTrigger] DurableOrchestrationContext context,
    TraceWriter log)
{
    // STEP 1: Get the list of blobs to process
    string[] dataFiles = await context.CallActivityAsync<string[]>(GetDataFilesFunction, "");

    // STEP 2 (FAN-OUT): Start multiple activity functions to process files in parallel.
    var parallelTasks = dataFiles.Select(
        file => context.CallActivityWithRetryAsync<Aggregates>(
            ProcessSingleFileFunction,
            GetRetryPolicy(),
            file));

    // STEP 3 (SLEEP): Wait for all the tasks to complete.
    await Task.WhenAll(parallelTasks);

    // STEP 4 (FAN-IN): Sleep until the parallel tasks are complete.
    int aggregateCount = 0;
    double aggregateSum = 0.0;
    double aggregateAverage = 0.0;
    foreach (var result in parallelTasks.Select(t => t.Result))
    {
        aggregateCount += result.Count;
        aggregateSum += result.Sum;
        aggregateAverage += result.Average;
    }

    aggregateAverage /= dataFiles.Length;

    log.Warning($"Results are in! Total ride count: {aggregateCount:N0}, total fees: ${aggregateSum:N2}, average cost: ${aggregateAverage:N2}.");

    // Return the aggregate results to the client.
    return new
    {
        Count = aggregateCount,
        Sum = aggregateSum,
        Average = aggregateAverage,
    };
}
```
The hang occurs on this line:
```csharp
foreach (var result in parallelTasks.Select(t => t.Result))
```
The problem is the call to `t.Result` inside `paralleltasks.Select`. Because `parallelTasks` is an enumerable, it will call `context.CallActivityAsync` each time it is enumerated. When called in the context of the `Select` method, a new task is created and that task has no chance to actually run and complete.

The fix to the above code is to add a `.ToArray()` to ensure the enumerable does not accidentally try to schedule new tasks.
```csharp
    var parallelTasks = dataFiles.Select(
        file => context.CallActivityWithRetryAsync<Aggregates>(
            ProcessSingleFileFunction,
            GetRetryPolicy(),
            file)).ToArray();
```
It's problematic to write code like this even in normal C# applications because the tasks will be run multiple times, but we should document this behavior since it may be non-obvious. Ideally there is a topic on what to look for if an orchestrator function hangs.
